sound/soc/sof/topology.c

```diff
@@ -2474,14 +2475,69 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
 	.bytes_ext_ops_count = ARRAY_SIZE(sof_dspless_bytes_ext_ops),
 };
 
+static const struct card_tplg_ops card_ops[] = {
+};
+
 int snd_sof_load_topology(struct snd_soc_component *scomp, const char *file)
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
 	const struct firmware *fw;
+	char **tplg_files;
+	int tplg_num = 0;
 	int ret;
+	int i;
 
 	dev_dbg(scomp->dev, "loading topology:%s\n", file);
 
+	for (i = 0; i < ARRAY_SIZE(card_ops); i++) {
+		if (!strcmp(scomp->card->name, card_ops[i].card_name)) {
+			ret = card_ops[i].get_tplg_files(scomp, file, &tplg_files);
+			if (ret < 0)
+				return ret;
+			tplg_num = ret;
+			break;
+		}
+	}
+	if (!tplg_num)
+		goto legacy_tplg;
+
+	for (i = 0; i < tplg_num; i++) {
+		dev_dbg(scomp->dev, "Requesting %d %s\n", i, tplg_files[i]);
+		ret = firmware_request_nowarn(&fw, tplg_files[i], scomp->dev);
+		if (ret < 0) {
+			if (i == 0) {
+				dev_dbg(scomp->dev, "Fail back to %s\n", file);
+				kfree(tplg_files[i]);
```

Comment:
why are we freeing this here, this function doesnt seem to allocate it?

Reply:
tplg_files[] are allocated in card_ops[i].get_tplg_files(). Moved after the `out:` label and added a comment to describe where they are allocated.